File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSame(a, b){
        if(!a && !b){
            return true;
        }
        if(a?.val != b?.val){
            return false;
        }
        return this.isSame(a.left, b.left) && this.isSame(a.right, b.right);

    }
    isSubtree(root, subRoot) {
        if(!subRoot){
            return true;
        }
        if(!root){
            return false;
        }
        // if(root.val == subRoot.val){
        //     return this.isSame(root, subRoot);
        // }



        return this.isSubtree(root.left, subRoot) || this.isSubtree(root.right, subRoot) || this.isSame(root, subRoot)
        
    }

}

File: Data Structures & Algorithms/house-robber-iii/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        cache = {}
        def dfs(node, available):
            if (node, available) in cache:
                return cache[(node, available)]
            if not node:
                return 0
            amount = 0
            if available:
                amount = node.val + dfs(node.left, False) + dfs(node.right, False)
            amount1 = dfs(node.left, True) + dfs(node.right, True)
            cache[(node, available)] = max(amount, amount1)
            return cache[(node, available)]
        return dfs(root, True)